File: typescript/src/loop-detector.ts
/**
 * Semantic loop detector that tracks generation history and detects
 * when the agents are repeating themselves without making progress.
 *
 * Uses two similarity strategies:
 * 1. Jaccard similarity (word-overlap) — fast, no external API needed.
 * 2. Cosine similarity over embeddings — more accurate, requires an embedding provider.
 */
export class SemanticLoopDetector {
  private history: string[] = [];
  private threshold: number;
  private windowSize: number;

  /**
   * @param threshold - Similarity score above which two texts are considered a loop (default 0.92).
   * @param windowSize - How many recent items to compare against (default 3).
   */
  constructor(threshold: number = 0.92, windowSize: number = 3) {
    this.threshold = threshold;
    this.windowSize = windowSize;
  }

  /**
   * Record a text in the history buffer.
   * Automatically prunes old entries beyond `windowSize * 2` to bound memory usage.
   */
  add(text: string): void {
    this.history.push(text);
    // Keep only recent history to bound memory
    if (this.history.length > this.windowSize * 2) {
      this.history = this.history.slice(-this.windowSize * 2);
    }
  }

  /**
   * Check whether `newText` is semantically too similar to any recent entry
   * using Jaccard (word-overlap) similarity.
   *
   * @returns true if a loop is detected.
   */
  isLooping(newText: string): boolean {
    const recentHistory = this.history.slice(-this.windowSize);
    for (const prev of recentHistory) {
      if (this.jaccardSimilarity(prev, newText) > this.threshold) {
        return true;
      }
    }
    return false;
  }

  /**
   * Check whether `newText` is semantically too similar to any recent entry
   * using cosine similarity over embeddings. More accurate but requires an
   * async embedding provider.
   *
   * @param newText - The new text to check.
   * @param getEmbedding - Async function that returns an embedding vector for a given text.
   * @returns true if a loop is detected.
   */
  async isLoopingWithEmbedding(
    newText: string,
    getEmbedding: (text: string) => Promise<number[]>,
  ): Promise<boolean> {
    const recentHistory = this.history.slice(-this.windowSize);
    if (recentHistory.length === 0) return false;

    const newEmb = await getEmbedding(newText);
    for (const prev of recentHistory) {
      const prevEmb = await getEmbedding(prev);
      if (this.cosineSimilarity(newEmb, prevEmb) > this.threshold) {
        return true;
      }
    }
    return false;
  }

  /**
   * Compute Jaccard similarity between two strings at the word level.
   * Ranges from 0 (no overlap) to 1 (identical word sets).
   */
  private jaccardSimilarity(a: string, b: string): number {
    const wordsA = new Set(a.toLowerCase().split(/\s+/));
    const wordsB = new Set(b.toLowerCase().split(/\s+/));
    const intersection = new Set([...wordsA].filter((x) => wordsB.has(x)));
    const union = new Set([...wordsA, ...wordsB]);
    if (union.size === 0) return 0;
    return intersection.size / union.size;
  }

  /**
   * Compute cosine similarity between two embedding vectors.
   * Ranges from -1 (opposite) to 1 (identical direction).
   */
  private cosineSimilarity(a: number[], b: number[]): number {
    let dot = 0;
    let normA = 0;
    let normB = 0;
    for (let i = 0; i < a.length; i++) {
      dot += a[i] * b[i];
      normA += a[i] * a[i];
      normB += b[i] * b[i];
    }
    const denom = Math.sqrt(normA) * Math.sqrt(normB);
    if (denom === 0) return 0;
    return dot / denom;
  }

  /**
   * Clear the entire history buffer.
   */
  reset(): void {
    this.history = [];
  }

  /**
   * Return the current number of entries in the history buffer.
   */
  get size(): number {
    return this.history.length;
  }
}

File: python/src/dual_agent_sdk/loop_detector.py
"""Semantic loop detection for the dual-agent orchestration SDK.

Provides :class:`SemanticLoopDetector`, a lightweight sliding-window detector
that flags when the Generator is producing near-identical output across
successive rounds.  Used by :class:`ConvergenceEngine` to trigger escalation.
"""

from __future__ import annotations

from collections import deque
from typing import Callable


class SemanticLoopDetector:
    """Detects semantic repetition in a stream of text strings.

    Maintains a fixed-size FIFO window of recent outputs.  Each time a new
    output arrives it is compared against every item in the window via Jaccard
    word-set similarity.  If any pairwise similarity exceeds *threshold* the
    output is considered a loop.

    When an embedding function is available, :meth:`check_with_embedding`
    provides a higher-quality cosine-similarity alternative.
    """

    # ------------------------------------------------------------------
    # Construction
    # ------------------------------------------------------------------

    def __init__(self, threshold: float = 0.92, window_size: int = 3) -> None:
        """Create a new loop detector.

        Args:
            threshold: Jaccard similarity above which two texts are
                considered semantically identical.  Range ``[0, 1]``.
            window_size: Maximum number of recent texts to retain for
                comparison.  Default is 3.
        """
        if not 0.0 <= threshold <= 1.0:
            raise ValueError("threshold must be between 0.0 and 1.0")
        if window_size < 1:
            raise ValueError("window_size must be at least 1")

        self.threshold = threshold
        self.window_size = window_size
        self._history: deque[str] = deque(maxlen=window_size)

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def add(self, text: str) -> None:
        """Append *text* to the comparison window.

        Should be called after each round's artifact is accepted or the
        verdict is recorded so that the next round can compare against it.
        """
        self._history.append(text)

    def is_looping(self, new_text: str) -> bool:
        """Check whether *new_text* is too similar to recent history.

        Returns:
            ``True`` if any item in the window has Jaccard similarity
            >= *threshold* with *new_text*.
        """
        if not self._history:
            return False

        tokens_new = self._tokenize(new_text)
        for past in self._history:
            if self._jaccard_similarity(tokens_new, self._tokenize(past)) >= self.threshold:
                return True
        return False

    async def check_with_embedding(
        self, new_text: str, get_embedding_fn: Callable[[str], list[float]]
    ) -> bool:
        """Loop check powered by cosine similarity of text embeddings.

        Useful when the caller has access to an embedding model (e.g.
        ``text-embedding-3-small``).  The function is expected to be
        a coroutine / async callable.

        Args:
            new_text: Candidate text to check.
            get_embedding_fn: Async callable that maps a string to a
                fixed-length embedding vector.

        Returns:
            ``True`` if the maximum cosine similarity with any window
            entry exceeds *threshold*.
        """
        if not self._history:
            return False

        emb_new = await get_embedding_fn(new_text)
        for past in self._history:
            emb_past = await get_embedding_fn(past)
            cos_sim = self._cosine_similarity(emb_new, emb_past)
            if cos_sim >= self.threshold:
                return True
        return False

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    @staticmethod
    def _tokenize(text: str) -> set[str]:
        """Lowercase whitespace tokenisation into a set."""
        return set(text.lower().split())

    @staticmethod
    def _jaccard_similarity(a: set[str], b: set[str]) -> float:
        """Jaccard coefficient of two token sets."""
        if not a and not b:
            return 1.0
        intersection = len(a & b)
        union = len(a | b)
        return intersection / union if union > 0 else 0.0

    @staticmethod
    def _cosine_similarity(a: list[float], b: list[float]) -> float:
        """Cosine similarity between two equal-length vectors."""
        dot = sum(x * y for x, y in zip(a, b))
        norm_a = sum(x * x for x in a) ** 0.5
        norm_b = sum(y * y for y in b) ** 0.5
        if norm_a == 0.0 or norm_b == 0.0:
            return 0.0
        return dot / (norm_a * norm_b)
